Remove old sdk from plugin

etanshaul opened the pull request:
fixes #781 

@patflynn, @joaoandremartins 

Removes the remaining references to the old SDK from the project.

- Remove appcfg from the facet editor
- Refactor all the helpers related to app engine standard (appengine-web.xml schema etc) into our CloudSdkService

Open issues:
- Determine how to support datanucleus (jdo/jpa) in the new plugin given that the libs are no longer bundled in the SDK. This support is broken by this PR, but will be fixed in a future one.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -40,4 +79,168 @@ public String getCloudSdkHomePath() {
+
+  @NotNull
+  @Override
+  public List<String> getUserLibraryPaths() {
```
This method should probably use the common library to get the SDK libraries. i.e., using getJarPath()

etanshaul replied:
it does pull it from gcloud. but since its not on the classpath I don't think we can use getJarPath?

joaoandremartins replied:
Since the libraries are under the Cloud SDK folder, in this case, we could add "appengine-endpoints" and "jsr107" to the keys returned by getJarPath(). We wouldn't need to assume that those jars are located at JAVA_TOOLS_BASE_PATH + "lib/opt/user", for example.

etanshaul replied:
Checkout my screenshots below (2nd to last one). Some of the lib's are actually directories. I wonder if we can use a similar mechanism to what you are suggesting.

joaoandremartins replied:
In that case, we could add extra methods to the library to get those directories too. I think the SDK clients shouldn't assume any knowledge of the internal SDK file system arrangement.

patflynn replied:
if this returns a list of paths should it just return List<Path> or List<File>?

etanshaul replied:
it does return a list? did you mean a list of Paths?

patflynn replied:
oh weird.  I guess my `<>` got removed from my comments by the markup formatting. my suggestion was `List<Path>` or `List<File>`

etanshaul commented:
For reference, here are the libraries that were previously configured in the bundled JB plugin:

![image](https://cloud.githubusercontent.com/assets/1735744/17710082/afff721a-63b9-11e6-805f-9216de866c64.png)

And here are the libraries configured when choosing datanucleus / jdo in the former bundled plugin:

![image](https://cloud.githubusercontent.com/assets/1735744/17710105/c8fa242c-63b9-11e6-8fd9-a08dfced7df6.png)

In the current state of our plugin, here is what is configured:
![image](https://cloud.githubusercontent.com/assets/1735744/17710118/d90755c4-63b9-11e6-8fe3-224af754a463.png)
jsr107 is missing

And in our plugin for persistence:
![image](https://cloud.githubusercontent.com/assets/1735744/17710129/eb176a6a-63b9-11e6-9c9b-2e0aca730164.png)
The libs are not resolved until we determine how to handle this in a future PR.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -16,21 +16,60 @@
+
   private PropertiesComponent propertiesComponent;
   private static final String CLOUD_SDK_PROPERTY_KEY = "GCT_CLOUD_SDK_HOME_PATH";
+  private static final String JAVA_TOOLS_BASE_PATH
```
Anything that assumes internal knowledge of the SDK should probably be better left to the library.
If you don't want to deal with removing this in this PR, should we add a TODO here or so?

etanshaul replied:
Yea. My suggestion would be to leave this here, open an issue against the lib, and then update this once its supported. Sound good to you?

etanshaul commented:
@patflynn / @joaoandremartins - any guesses as to why [jsr107](https://jcp.org/en/jsr/detail?id=107) would have been added as a lib by the old plugin? This is not bundled in gcloud so I'm wondering if we want or need to bring this in.

joaoandremartins commented:
Not sure. @ludoch might know, since I recall this being added to Eclipse as well.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/descriptor/AppEngineWebSchemaProvider.java:
```diff
@@ -73,10 +74,14 @@ public XmlFile getSchema(@NotNull @NonNls String url, @Nullable Module module,
+          file = CloudSdkService.getInstance().getApplicationSchemeFile();
         } else {
-          file = facet.getSdk().getWebSchemeFile();
+          file = CloudSdkService.getInstance().getWebSchemeFile();
```
what's the difference between these two files? (web vs application scheme)

etanshaul replied:
I was wondering that too. Maybe appengine-application.xml is an old concept? This is what the template looks like:

```
<appengine-application xmlns="http://appengine.google.com/ns/1.0">
  <application>app</application>
</appengine-application>
```

patflynn replied:
ah I think I remember what this is. It's for multi-module/service java apps that are bundled as EARs.  Only appcfg supports it so it probably can be safely removed. People might be a little surprised when they find that deployment doesn't chain all their separate services..

@ludoch

etanshaul replied:
Yea, we currently do not support ear deployments. The individual modules need to be deployed separately so I agree this can be removed if that's what it was for.

patflynn replied:
let's remove the ApplicationSchemeFile handling.

etanshaul replied:
Done - see https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/822/commits/fae3ae03423fb53ba4de4862dc85bf063e37a7e2

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFrameworkDetector.java:
```diff
@@ -56,10 +52,6 @@ public FrameworkType getFrameworkType() {
 
   @Override
   public void setupFacet(@NotNull AppEngineFacet facet, ModifiableRootModel model) {
```
does this have to be implemented  if it's a no-op?

etanshaul replied:
it doesn't. I'll remove it.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineSupportProvider.java:
```diff
@@ -150,9 +146,10 @@ private void addSupport(final Module module,
+    webIntegration.addDevServerToModuleDependencies(rootModel);
 
-    final Library apiJar = addProjectLibrary(module, "AppEngine API", sdk.getUserLibraryPaths(),
+    final Library apiJar = addProjectLibrary(module, "AppEngine API",
```
So does this happen once on creating the facet?

etanshaul replied:
Yes, but I believe the trigger is adding the framework support (which in turn creates the facet).

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/impl/AppEngineCommunityWebIntegration.java:
```diff
@@ -82,18 +81,15 @@ public void setupJpaSupport(@NotNull Module module, @NotNull VirtualFile persist
-  public void setupRunConfiguration(@NotNull AppEngineSdk sdk,
-      @Nullable Artifact artifact,
-      @NotNull Project project) {
+  public void setupRunConfiguration(@Nullable Artifact artifact, @NotNull Project project) {
```
why is this class called ..Community..?  (Lots of questions from ignorance here. sorry)

etanshaul replied:
So the way it works is that plugin.xml in our main cloud-tools module defines a WebIntegration inteface. The same config file then injects a Community implementation for that module.

Similarly, the ultimate module injects an overriding Ultimate implementation. This module is only loaded when running UE.

The CE version is basically just a noop implementation.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkService.java:
```diff
@@ -27,7 +35,44 @@ public static CloudSdkService getInstance() {
   public abstract String getCloudSdkHomePath();
 
   public abstract void setCloudSdkHomePath(String path);
+
```
a lot of this kind of stuff does kind of feel like it could be passed straight through to the common lib. Wrapping it as an IJ service isn't terrible though, but then the name of the service seems a bit off.

etanshaul replied:
By passing through to the common lib do you mean having the common lib expose this functionality for us? That is not a bad idea as I could see this being useful to other clients like ecplise.

etanshaul replied:
And if I interpreted you correctly, how about we leave it in the service and open an enhancement against the common lib to expose this functionality for us.

I'll think about a better name for the service since its doing a bunch of appengine standard'y stuff now.

patflynn replied:
that works.

etanshaul commented:
@patflynn, @joaoandremartins ptal

In my last commit I cleaned up the file path related logic for the sdk and the AE standard libs. I'm going to look into test coverage, but you can review in the meantime.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkService.java:
```diff
@@ -27,7 +35,44 @@ public static CloudSdkService getInstance() {
   public abstract void setCloudSdkHomePath(String path);
+
+  @Nullable
+  public abstract File getToolsApiJarFile();
```
I'm assuming these are still Files to fit more easily in the IJ framework?

etanshaul replied:
Yea, thats right.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -40,4 +81,214 @@ public String getCloudSdkHomePath() {
+        && getToolsApiJarFile().exists();
+  }
+
+  private static String findLatestVersion(File dir) {
```
Do we still need this if we're not supporting SDK versioning?

etanshaul replied:
Actually, the consumers of this are using it to get the latest version of the endpoints and jsr107 libs.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -16,21 +16,62 @@
+
   private PropertiesComponent propertiesComponent;
   private static final String CLOUD_SDK_PROPERTY_KEY = "GCT_CLOUD_SDK_HOME_PATH";
+  private static final Path JAVA_TOOLS_RELATIVE_PATH
```
Add TODO to get rid this by adding to the library.

etanshaul replied:
Will do.

patflynn commented:
I'm looking at this now. Sorry for the delay.

patflynn commented on google-cloud-tools-plugin/jps-plugin/src/com/google/cloud/tools/intellij/appengine/jps/model/impl/JpsAppEngineModuleExtensionImpl.java:
```diff
@@ -74,7 +77,7 @@ public void applyChanges(@NotNull JpsAppEngineModuleExtensionImpl modified) {
   public String getToolsApiJarPath() {
     return FileUtil.toSystemDependentName(
-        JpsArtifactPathUtil.appendToPath(getSdkHomePath(), LIB_APPENGINE_TOOLS_API_JAR));
+        JpsArtifactPathUtil.appendToPath(getSdkHomePath(), LIB_APPENGINE_TOOLS_API_JAR.toString()));
```
https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/186

patflynn replied:
maybe a todo to update once the issue in link is closed?

etanshaul replied:
Yea definitely. This is inline with @joaoandremartins's suggestions to mark todos to offload things to the common lib where it makes sense.

patflynn replied:
still don't see a TODO: here

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineSupportProvider.java:
```diff
@@ -187,7 +184,8 @@ private void addSupport(final Module module,
       }
       final Library library = addProjectLibrary(module, "AppEngine ORM",
-          Collections.singletonList(sdk.getOrmLibDirectoryPath()), sdk.getOrmLibSources());
+          Collections.singletonList(CloudSdkService.getInstance().getOrmLibDirectoryPath()),
```
not a biggie, but unless I was worried that CloudSdkService could change mid thread, I'd be tempted to just extract a field or variable and reuse it across invocations.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -40,4 +83,206 @@ public String getCloudSdkHomePath() {
   }
+
+  @Nullable
+  private Path getJavaToolsBasePath() {
```
not tempted to just blow up in these null cases?

etanshaul replied:
my reasoning was that I didn't want the IDE to explode if gcloud ever changes the way it packages these libs. But maybe having it explode would be a way for us to quickly be informed. What do you think?

patflynn replied:
let's leave the nullable returns as is then. I don't think logging.error makes sense since as you say, these methods can be called when the user doesn't have an SDK set.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -40,4 +83,206 @@ public String getCloudSdkHomePath() {
+  @Override
+  // TODO this path is incorrect. We need to determine an alternate strategy for loading these libs
+  // maybe from maven central
+  public VirtualFile[] getOrmLibSources() {
```
do we need any of these since we're not doing jpa/jdo support?

etanshaul replied:
no. but I'd like to remove all of this in a different PR since this one is already a little big.

patflynn replied:
Yeah if this is erroring I'd want to know about it

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -40,4 +83,206 @@ public String getCloudSdkHomePath() {
+      result.add(libUserDirectory);
+    }
+
+    if (getJavaToolsBasePath() == null) {
```
doesn't libUserDirectory depend on getJavaToolsBasePath not being null?

etanshaul replied:
If javaToolsBasePath is null it will return an empty list which was just to satisfy the caller that will NPE otherwise. But I suppose this should be refactored.

patflynn replied:
what I meant is maybe this should be the first condition.  i.e.  shortcircuit if getJavaToolsBasPath is null, then add the libUserDirectory, since you know it exists by that point.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -40,4 +83,206 @@ public String getCloudSdkHomePath() {
+    return null;
+  }
+
+  private Map<String, Set<String>> loadBlackList() throws IOException {
```
do you know if anybody has file an issue in appengine-plugins-core to provide an API for this?

etanshaul replied:
no I don't think so. I can file one though.

etanshaul replied:
do you know where this method blacklist even comes from? It was pulled from old plugin, but I don't see this documented anywhere in AE (whereas the class whitelist is documented).

patflynn replied:
wait wasn't this replaced by the whitelist?

On Thu, Aug 18, 2016 at 3:10 PM, Etan Shaul <notifications@github.com>
wrote:

> In google-cloud-tools-plugin/src/com/google/cloud/tools/
> intellij/appengine/sdk/DefaultCloudSdkService.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/822#discussion_r75370270>
> :
>
> > +  public boolean isValid() {
> > +    return !StringUtil.isEmpty(getCloudSdkHomePath())
> > +        && getToolsApiJarFile() != null
> > +        && getToolsApiJarFile().exists();
> > +  }
> > +
> > +  private static String findLatestVersion(File dir) {
> > +    String[] names = dir.list();
> > +    if (names != null && names.length > 0) {
> > +      String max = Collections.max(Arrays.asList(names));
> > +      return FileUtil.toSystemIndependentName(new File(dir, max).getAbsolutePath());
> > +    }
> > +    return null;
> > +  }
> > +
> > +  private Map<String, Set<String>> loadBlackList() throws IOException {
>
> do you know where this method blacklist even comes from? It was pulled
> from old plugin, but I don't see this documented anywhere in AE (whereas
> the class whitelist is documented).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/822/files/0bbf4e19de562c3c734710e081ac539e182b4686#r75370270>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HVnuJ9nOTU704gdr0w1Oj5dee16dks5qhK4WgaJpZM4Jlk8z>
> .
>

etanshaul replied:
No. My original explanation of all this was incorrect:

There are two separate concepts, a _class_ whitelist (which the common lib is providing us and was previously pulled from the old sdk jar), and a _method_ blacklist which blacklists some methods on classes in the whilelist (for example `System` is whitelisted, but `System#exit` is blacklisted).

Its this blacklist that I don't know where JB got it from. It's a plain text file in the resources folder.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java:
```diff
@@ -40,4 +83,206 @@ public String getCloudSdkHomePath() {
+
+  private static File[] getJarsFromDirectory(File libFolder) {
+    List<File> jars = new ArrayList<>();
+    final File[] files = libFolder.listFiles();
```
I think this does I/O.  Can we make this method non-static?

patflynn commented on google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/integration/AppEngineServerHelper.java:
```diff
@@ -32,14 +33,14 @@
   }
 
   public ApplicationServerPersistentData createPersistentDataEmptyInstance() {
-    return new AppEngineServerData("");
```
just curious about this change. What is it even doing? how does the dta object get used?

etanshaul replied:
Good point. The config subclass (AppEngineServerData) was removed. This whole class can actually be deleted too as its not doing anything.

patflynn commented on google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/run/AppEngineServerStartupPolicy.java:
```diff
@@ -38,17 +37,14 @@
-    if (StringUtil.isEmpty(sdk.getSdkHomePath())) {
+    final CloudSdkService sdkService = CloudSdkService.getInstance();
+    if (StringUtil.isEmpty(sdkService.getCloudSdkHomePath())) {
       throw new ExecutionException("Path to App Engine SDK isn't specified");
```
reword error message from App Engine SDK to Cloud SDK

etanshaul replied:
this class gets removed by https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/811 so I didn't bother

patflynn replied:
ok

patflynn commented on google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/run/AppEngineServerStartupPolicy.java:
```diff
@@ -38,17 +37,14 @@
       throw new ExecutionException(
-          "'" + sdk.getSdkHomePath() + "' isn't valid App Engine SDK installation: '"
-              + toolsApiJarFile.getAbsolutePath() + "' not found");
+          "'" + sdkService.getCloudSdkHomePath() + "' isn't valid App Engine SDK installation: '");
```
same here

etanshaul commented:
@patflynn / @joaoandremartins I think I've address all of your comments and cleaned things up a bit in my last several commits. ptal.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/AppEngineCloudSdkService.java:
```diff
@@ -0,0 +1,290 @@
+  private Map<String, Set<String>> myMethodsBlackList;
+
+  public AppEngineCloudSdkService() {
+    this.propertiesComponent = PropertiesComponent.getInstance();
```
You could initialize propertiesComponent at declaration time, since you're not assigning anywhere else.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/AppEngineCloudSdkService.java:
```diff
@@ -0,0 +1,290 @@
+
+  @Nullable
+  @Override
+  public String getSdkHomePath() {
```
FYI I'm changing this in my PR to return a Path.

joaoandremartins commented:
Aside from the propertiesComponent nit I pointed out, LGTM.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/AppEngineCloudSdkService.java:
```diff
@@ -0,0 +1,290 @@
+ * serialization.
+ *
+ */
+// TODO Offload various path logic for retrieving AE libs to the common library once implemented
```
for TODO comments I'm not sure what our github strategy but I'd be fine with "TODO({github-user-name}): ..." which would match our styleguide.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/AppEngineCloudSdkService.java:
```diff
@@ -0,0 +1,290 @@
+ *
+ */
+// TODO Offload various path logic for retrieving AE libs to the common library once implemented
+public class AppEngineCloudSdkService extends AppEngineSdkService {
```
This renaming has surprised me.  The concept of an 'App Engine SDK' seems to be going away. But let's not churn on it. This is fine and we can think about the right way to model cloud sdk vs App Engine tools at a later date.

patflynn commented:
Feel free to merge once you address my last few comments.

etanshaul commented:
I addressed the last couple of comments:

- I renamed the SDK service back to what it was (I agree the 'app engine sdk' wording isn't the best)
- Added a missing todo and added username to them
- Initialize member var at declaration

I'll merge this in once the build is green.
